compressor/compressor.cpp

```diff
@@ -0,0 +1,165 @@
+{
+  int start = 0;
+  bool checking = false;
+
+  for (unsigned int k = 0; k < i; ++k)
+  {
+    if (window[k] == window[i] && checking == false)
+    {
+      checking = true;
+      start = k;
+    }
+    else if (checking == true && (k - start == max_match_length ||window.size() == i + k - start || window[k] != window[i + k - start]))
+    {
+      checking = false;
+      if ((unsigned int)(off_len.get_length()) < k - start)
+        off_len.set(i - start, k - start);
+
+      if (off_len.get_length() == max_match_length)
+        break;      
+
+      k = start + 1;
+      start = 0;
+    }
+  }
+
+  if (off_len.get_offset() == 0) put_error_and_exit("Offset is impossible\n");//
+  if (off_len.get_length() <= 0 || off_len.get_length() > max_match_length) put_error_and_exit("Length is impossible\n");//
+
+  if (off_len.get_length() > 1)
+    return true;
+  else
+    return false;
+}
+
+void compressor(const std::string& filename)
+{
+  std::ifstream in (filename, std::ios::in|std::ios::binary|std::ios::ate);
+  std::ofstream out (filename + ".lzw", std::ios::out|std::ios::binary|std::ios::trunc);
+  std::streampos end_of_file = in.tellg();
+  in.seekg(0, std::ios::beg);
+
+  std::deque<Byte> window;
+  BitCashier bits;
+  
+  for (unsigned int i = 0; i < window_width && in.tellg() < end_of_file; ++i)
+    window.push_back(in.get());
+
+  for (unsigned int i = 0; i < window.size(); ++i)
+  {
+    i = slide_window(i, window, in, end_of_file);
+    OffsetAndLength off_len;
+
+    if (find_a_match(window, i, off_len))
+    {
+      bits.add(true);
+      //std::cout << "<" << off_len.get_offset() << "," << off_len.get_length() << ">";//
+      bits.add(off_len.get_offset(true));
+      bits.add(off_len.get_length(true), 4);
+      i += off_len.get_length() - 1;
+    }
+    else
+    {
+      bits.add(false);
+      //std::cout << char(window[i]);//
+      bits.add((window[i]));
+    }
+
+    while (bits.size() >= byte_length)
+      out.put(bits.pop_bits());
+  }
+
+  while (bits.size() > 0 && bits.size() < 8)
+    bits.add(false);
+
+  if (bits.size() == 8)
+    out.put(bits.pop_bits());
+}
+
+void decompressor(const std::string& filename)
+{
+  std::ifstream in (filename, std::ios::in|std::ios::binary|std::ios::ate);
+  std::ofstream out (filename.substr(0, filename.length() - 4), std::ios::out|std::ios::binary|std::ios::trunc);
+  std::streampos end_of_file = in.tellg();
+  in.seekg(0, std::ios::beg);
+
+  std::deque<Byte> window;
+  BitCashier bits;
+
+  while (in.tellg() < end_of_file)
+  {
+    bits.add(Byte(in.get()));
+    if(bits.read_not_pop() && bits.size() >= encoded_length)
+    {
+      bits.pop();
+      OffsetAndLength off_len;
+      off_len.set_offset(bits.pop_bits(), true);
+      off_len.set_length(bits.pop_bits(4), true);
+      //std::cerr << "<" << off_len.get_offset() << ", " << off_len.get_length() << ">";//
+
+      for (unsigned int i = 0; i < (unsigned int)(off_len.get_length()); ++i)
```

Comment:
If you use "unsigned" you can write unsigned(off_len.get_length()) which is much better.